src/tools/dashboard/src/reference.rs

```diff
@@ -0,0 +1,287 @@
+    // build configs. `x.py` respects those configs, causing a recompilation
+    // of `rustc`. This is not a desired behavior, so we remove those configs.
+    let filtered_env: HashMap<String, String> = env::vars()
+        .filter(|&(ref k, _)| {
+            !(k.contains("CARGO") || k.contains("LD_LIBRARY_PATH") || k.contains("RUST"))
+        })
+        .collect();
+    let mut cmd = Command::new([".", "x.py"].iter().collect::<PathBuf>());
+    cmd.args([
+        "test",
+        suite,
+        "-i",
+        "--stage",
+        "1",
+        "--test-args",
+        "--logfile",
+        "--test-args",
+        log_path.to_str().unwrap(),
+    ]);
+    cmd.env_clear().envs(filtered_env);
+    cmd.stdout(Stdio::null());
+
+    cmd.spawn()
+        .expect(&format!("Error: failed to spawn process:\n{:?}", cmd))
+        .wait()
+        .expect(&format!("Error: spawned process did not run:\n{:?}", cmd));
+}
+
+/// Creates a new [`Tree`] from a path, `ns`, and a test result, `p`.
+fn tree_from_path(mut ns: Vec<String>, p: bool) -> dashboard::Tree {
+    assert!(ns.len() > 0, "Error: `ns` must contain at least 1 element.");
+    let mut tree = dashboard::Tree::new(
+        dashboard::Node::new(ns.pop().unwrap(), if p { 1 } else { 0 }, if p { 0 } else { 1 }),
+        vec![],
+    );
+    for _ in 0..ns.len() {
+        tree = dashboard::Tree {
+            data: dashboard::Node {
+                name: ns.pop().unwrap(),
+                num_pass: tree.data.num_pass,
+                num_fail: tree.data.num_fail,
+            },
+            children: vec![tree],
+        };
+    }
+    tree
+}
+
+/// Parses and generates a dashboard from the log output of `compiletest` in
+/// `path`.
+fn parse_log(path: &Path) -> dashboard::Tree {
+    let file = fs::File::open(path).expect(&format!("Error: failed to open log file: {:?}", path));
+    let reader = BufReader::new(file);
+    let mut tests = dashboard::Tree {
+        data: dashboard::Node { name: String::from("ref"), num_pass: 0, num_fail: 0 },
+        children: vec![],
+    };
+    for line in reader.lines() {
+        let (ns, l) = parse_log_line(&line.unwrap());
+        tests = dashboard::Tree::merge(tests, tree_from_path(ns, l)).unwrap();
+    }
+    tests
+}
+
+/// Parses a line in the log output of `compiletest` and returns a pair containing
+/// the path to a test and its result.
+fn parse_log_line(line: &str) -> (Vec<String>, bool) {
+    // Each line has the format `<result> [rmc] <path>`. Extract <result> and
+    // <path>.
+    let splits: Vec<_> = line.split(" [rmc] ").map(String::from).collect();
+    let l = match splits[0].as_str() {
+        "ok" => true,
+        _ => false,
+    };
+    let mut ns: Vec<_> = splits[1].split(&['/', '.'][..]).map(String::from).collect();
+    // Remove unnecessary `.rs` suffix.
+    ns.pop();
+    (ns, l)
+}
+
+/// Display the dashboard in the terminal.
+fn display_dashboard(dashboard: dashboard::Tree) {
```

Comment:
If our preprocessing allows us to inject assertions into the tests then is there a way we can distinguish between:
- codegen failure
- codegen success + verification fail
- codegen success + verification success
?

Reply:
Yes. We can add a new step that runs `rmc-rustc` on the extracted code (after pre-processing?) and save the result in a new field in the `Node` struct. I wonder if I should do that through `compiletest`...